### PR TITLE
Compile audio, editor, renderer

### PR DESCRIPTION
**Pull request by MAnyKey**

### CODE REVIEW

File: src/editor/profiler_ui.cpp [hunk 1]
```diff
@@ -378,8 +378,8 @@ struct ProfilerUIImpl : public ProfilerUI
 		Lumix::Array<float> m_frames;
 		struct Hit
 		{
-			Lumix::uint64 start;
```
Comment (nem0) [inline src/editor/profiler_ui.cpp]: why was this changed?

File: src/renderer/renderer.cpp [hunk 2]
```diff
@@ -752,14 +752,13 @@ struct RendererImpl : public Renderer
...
-			if (compareString(m_passes[i], pass) == 0)
+			if (m_passes[i] == pass)
 			{
 				return i;
 			}
 		}
 
-		auto& new_pass = m_passes.emplace();
-		copyString(new_pass, pass);
+		auto& new_pass = m_passes.emplace(pass);
```
Comment (nem0) [inline src/renderer/renderer.cpp]: ```auto& new_pass = ``` is not necessary

File: src/renderer/shader.cpp [hunk 3]
```diff
@@ -260,7 +260,7 @@ static void uniform(lua_State* L, const char* name, const char* type)
 static void pass(lua_State* state, const char* name)
 {
 	auto* cmb = getCombinations(state);
-	copyString(cmb->m_passes[cmb->m_pass_count], name);
+	copyString(cmb->m_passes[cmb->m_pass_count].data, name);
```
Comment (nem0) [inline src/renderer/shader.cpp]: Just curious, does this fail to compile on Linux? What is the error message

File: src/audio/clip_manager.cpp [hunk 4]
```diff
@@ -4,7 +4,7 @@
 #include "engine/core/string.h"
 #include "engine/lumix.h"
 #define STB_VORBIS_HEADER_ONLY
-#include "stb_vorbis.c"
+#include "stb_vorbis.h"
```
Comment (nem0) [inline src/audio/clip_manager.cpp]: see appveyor

File: src/editor/metadata.h [hunk 5]
```diff
@@ -42,7 +42,7 @@ class LUMIX_EDITOR_API Metadata
 			int m_int;
 			char m_string[Lumix::MAX_PATH_LENGTH];
 			
-			struct Raw
+			struct
```
Comment (nem0) [inline src/editor/metadata.h]: https://msdn.microsoft.com/en-us/library/c89bw853.aspx

Comment (nem0) [inline src/editor/metadata.h]: Never mind, it's ok, diff viewer cut-off one line

File: src/editor/profiler_ui.cpp (hunk 1 above)
Comment (MAnyKey) [inline src/editor/profiler_ui.cpp]: pointer to hit contents is used [here](https://github.com/MAnyKey/LumixEngine/blob/linux-port-9/src/editor/profiler_ui.cpp#L692) btw I just realized that better decision is to allocate temporary array right there and pass pointer to it.

I changed definition of `uint64` to `unsigned long` on 64-bit Linux so that it is compatible with `cstdint` (otherwise there is bunch of error messages from usages of bgfx which uses `cstdint`'s ints.

File: src/renderer/shader.cpp (hunk 3 above)
Comment (MAnyKey) [inline src/renderer/shader.cpp]: It is a consequence of changing Pass from `char[20]` to `StaticString<20>`. `copyString` takes static-sized C array or pointer and length.
Original errors were located in `array.h` when using `Array<Pass>` like that:

```
In file included from ../../../src/renderer/renderer.cpp:3:0:
../../../src/engine/core/array.h: In instantiation of ‘T& Lumix::Array<T>::emplace(Params&& ...) [with Params = {const char*&}; T = char [32]]’:
../../../src/renderer/renderer.cpp:746:34:   required from here
../../../src/engine/core/array.h:222:3: error: no matching function for call to ‘operator new [](sizetype, Lumix::NewPlaceholder, char*)’
   new (NewPlaceholder(), (char*)(m_data + m_size)) T(myforward<Params>(params)...);
   ^
../../../src/engine/core/array.h:222:3: note: candidate is:
<built-in>:0:0: note: void* operator new [](long unsigned int)
<built-in>:0:0: note:   candidate expects 1 argument, 3 provided
../../../src/engine/core/array.h: In instantiation of ‘void Lumix::Array<T>::callDestructors(T*, T*) [with T = char [32]]’:
../../../src/engine/core/array.h:90:42:   required from ‘Lumix::Array<T>::~Array() [with T = char [32]]’
../../../src/renderer/renderer.cpp:623:26:   required from here
../../../src/engine/core/array.h:364:4: error: request for member ‘~char [32]’ in ‘* begin’, which is of non-class type ‘char [32]’
    begin->~T();
    ^

```

Comment (nem0) [inline src/renderer/shader.cpp]: I was suprised why StaticString was not implicitly cast to const char*, but now I understand.

File: src/editor/profiler_ui.cpp (hunk 1 above)
Comment (nem0) [inline src/editor/profiler_ui.cpp]: Yeah, I do not like that bgfx is "leaking" in engine, but it would require a lot of boilerplate code to abstract it. By the way IntervalGraph is LumixEngine's custom code created and used only for this purpose, so it can be changed if necessary. 

> (otherwise there is bunch of error messages from usages of bgfx which uses cstdint's ints.

Was is something about unsigned long to unsigned long long conversion? 

I do not like that [there is a little type mismatch now.](https://github.com/nem0/LumixEngine/blob/master/src/editor/profiler_ui.cpp#L490)

File: src/audio/clip_manager.cpp [hunk 6]
```diff
@@ -4,7 +4,7 @@
 #include "engine/core/string.h"
 #include "engine/lumix.h"
 #define STB_VORBIS_HEADER_ONLY
-#include "stb_vorbis.h"
+#include "stb_vorbis.cpp"
```
Comment (nem0) [inline src/audio/clip_manager.cpp]: This will make it compile with C++ compiler instead of C on Windows, hopefully nothing will break

File: src/editor/profiler_ui.cpp (hunk 1 above)
Comment (MAnyKey) [inline src/editor/profiler_ui.cpp]: > Was is something about unsigned long to unsigned long long conversion?

Yeah, something like that. bgfx has [interface](https://github.com/nem0/LumixEngine/blob/master/external/bgfx/include/bgfx/bgfx.h#L315) where some methods take `uint64_t` as parameter. Lumix's code cannot override it with it's own `uint64`.

**Comment by nem0** [comment]
#655
